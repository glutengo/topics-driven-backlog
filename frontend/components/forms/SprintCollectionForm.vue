<template>
    <BForm 
        @submit.prevent="submit" 
        class="sprint-collection-form">

        <BRow>
            <BCol md="6">
                <h4>Add multiple sprints</h4>
                <p>
                    This feature allows you to add multiple sprints at the same time. Just enter a sprint length in days
                    (for example 7 for week-long sprints) and the start and end date of the period you'd like to generate sprints for.
                    You can view a preview of the sprints that are going to be generated for you on the right.
                    Please note that you can only add future sprints with this feature.
                </p>

                <BFormGroup 
                    label="Sprint length in days" 
                    label-for="sprint-collection-duration">

                    <BFormInput 
                        id="sprint-collection-duration" 
                        type="number" 
                        min="1" 
                        step="1" 
                        v-model="data.duration" 
                        required/>
                </BFormGroup>

                <BFormGroup 
                    label="Start date" 
                    label-for="sprint-collection-start-date">

                    <BFormInput 
                        id="sprint-collection-start-date" 
                        :min="$store.state.currentDate" 
                        type="date" 
                        v-model="data.start_date" 
                        required/>
                </BFormGroup>

                <BFormGroup 
                    label="End date" 
                    label-for="sprint-collection-end-date">

                    <BFormInput 
                        id="sprint-collection-end-date" 
                        :min="data.start_date" 
                        type="date" 
                        v-model="data.end_date" 
                        required/>
                </BFormGroup>

                <BButton 
                    type="submit" 
                    variant="primary" 
                    :disabled="isInvalid">Add sprints</BButton>

                <BButton 
                    type="button" 
                    variant="grey" 
                    @click="cancel">Cancel</BButton>
            </BCol>
            <BCol md="6">
                <h4>Preview</h4>
                <p v-if="isInvalid">
                    Please fill out all fields correctly to generate the preview.
                </p>

                <p v-else-if="numberOfSprints < 1">
                    We cannot fit any sprints of the given duration into the period you provided.
                </p>

                <template v-else>
                    <div 
                        v-for="i in numberOfSprints" 
                        :key="i">

                        <strong>{{ i }}. Sprint:</strong>

                        {{ datePlusDays(data.start_date, (i - 1) * data.duration) }}
                        -
                        {{ datePlusDays(data.start_date, i * data.duration - 1) }}
                    </div>
                </template>
            </BCol>
        </BRow>
    </BForm>
</template>

<script>
import moment from 'moment';
import { required } from 'vuelidate/lib/validators';
import BForm from '@bootstrap/form/form';
import BFormGroup from '@bootstrap/form-group/form-group';
import BFormInput from '@bootstrap/form-input/form-input';
import BRow from '@bootstrap/layout/row';
import BCol from '@bootstrap/layout/col';
import BButton from '@bootstrap/button/button';
import smallerOrEqualThan from '@/validators/smallerOrEqualThan';
import largerOrEqualThan from '@/validators/largerOrEqualThan';
import BaseForm from '@/components/forms/BaseForm';

export default {
    name: 'SprintCollectionForm',
    components: { BForm, BFormGroup, BFormInput, BRow, BCol, BButton },
    extends: BaseForm,
    computed: {
        /**
         * Number of days between start and end date in data
         * @returns {int}
         */
        numberOfDays() {
            if (this.isInvalid) {
                return -1;
            }

            const start = moment(this.data.start_date, 'YYYY-MM-DD');
            const end = moment(this.data.end_date, 'YYYY-MM-DD');

            // add 1 since both start and end date are included
            return end.diff(start, 'days') + 1;
        },

        /**
         * Number of sprints that fit into the number of days between dates
         * @returns {int}
         */
        numberOfSprints() {
            return this.numberOfDays >= 0 ? Math.floor(this.numberOfDays / this.data.duration) : 0;
        },
    },
    methods: {
        /**
         * Add a number of days to a given date
         * @param {string} date
         * @param {int} days
         * @returns {string}
         */
        datePlusDays(date, days) {
            return moment(date, 'YYYY-MM-DD')
                .add(days, 'days')
                .format('YYYY-MM-DD');
        },
    },
    validations: {
        data: {
            duration: { required },
            start_date: {
                required,
                smallerThanEnd: smallerOrEqualThan('end_date'),
                largerThanToday: largerOrEqualThan(moment().format('YYYY-MM-DD'), true),
            },
            end_date: {
                required,
                largerThanStart: largerOrEqualThan('start_date'),
            },
        },
    },
};
</script>