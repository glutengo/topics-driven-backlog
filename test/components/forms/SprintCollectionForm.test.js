import { mount, createLocalVue } from 'vue-test-utils';
import moment from 'moment';
import Vuex from 'vuex';
import Vuelidate from 'vuelidate';
import SprintCollectionForm from '@/components/forms/SprintCollectionForm';

// take a date far in the future so validations pass
Date.now = jest.fn(() => 7956915742000);

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Vuelidate);

describe('SprintCollectionForm.test.js', () => {
    let cmp, store;

    beforeEach(() => {
        store = new Vuex.Store({
            currentDate: moment().format('YYYY-MM-DD'),
        });

        cmp = mount(SprintCollectionForm, {
            localVue,
            store,
            propsData: {
                data: {
                    duration: 2,
                    start_date: moment().format('YYYY-MM-DD'),
                    end_date: moment()
                        .add(5, 'days')
                        .format('YYYY-MM-DD'),
                },
            },
        });
    });

    it('calculates the correct number of days', () => {
        expect(cmp.vm.numberOfDays).toBe(6);
    });

    it('calculates the correct number of sprints', () => {
        expect(cmp.vm.numberOfSprints).toBe(3);
    });

    it('doesnt do calculations when form is invalid', () => {
        cmp.setProps({
            data: {
                duration: 2,
                start_date: moment().format('YYYY-MM-DD'),
                end_date: moment()
                    .subtract(1, 'days')
                    .format('YYYY-MM-DD'),
            },
        });

        expect(cmp.vm.numberOfDays).toBe(-1);
        expect(cmp.vm.numberOfSprints).toBe(0);
    });

    it('adds days to a date', () => {
        expect(cmp.vm.datePlusDays('2000-01-01', 1)).toBe('2000-01-02');
    });
});
