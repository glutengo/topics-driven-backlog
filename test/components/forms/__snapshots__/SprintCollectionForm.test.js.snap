// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SprintCollectionForm.test.js has the expected html structure 1`] = `
<form
  class="sprint-collection-form"
>
  <div
    class="row"
  >
    <div
      class="col-md-6"
    >
      <h4>
        Add multiple sprints
      </h4>
       
      <p>
        
                This feature allows you to add multiple sprints at the same time. Just enter a sprint length in days
                (for example 7 for week-long sprints) and the start and end date of the period you'd like to generate sprints for.
                You can view a preview of the sprints that are going to be generated for you on the right.
                Please note that you can only add future sprints with this feature.
            
      </p>
       
      <div
        aria-labelledby="__BVID__99___BV_label_"
        class="b-form-group form-group"
        id="__BVID__99_"
        role="group"
      >
        <label
          class="col-form-label pt-0"
          for="sprint-collection-duration"
          id="__BVID__99___BV_label_"
        >
          Sprint length in days
        </label>
        <div
          class=""
        >
          <input
            aria-required="true"
            class="form-control"
            id="sprint-collection-duration"
            min="1"
            required="required"
            step="1"
            type="number"
          />
          <!---->
          <!---->
          <!---->
        </div>
      </div>
       
      <div
        aria-labelledby="__BVID__101___BV_label_"
        class="b-form-group form-group"
        id="__BVID__101_"
        role="group"
      >
        <label
          class="col-form-label pt-0"
          for="sprint-collection-start-date"
          id="__BVID__101___BV_label_"
        >
          Start date
        </label>
        <div
          class=""
        >
          <input
            aria-required="true"
            class="form-control"
            id="sprint-collection-start-date"
            required="required"
            type="date"
          />
          <!---->
          <!---->
          <!---->
        </div>
      </div>
       
      <div
        aria-labelledby="__BVID__103___BV_label_"
        class="b-form-group form-group"
        id="__BVID__103_"
        role="group"
      >
        <label
          class="col-form-label pt-0"
          for="sprint-collection-end-date"
          id="__BVID__103___BV_label_"
        >
          End date
        </label>
        <div
          class=""
        >
          <input
            aria-required="true"
            class="form-control"
            id="sprint-collection-end-date"
            min="2222-02-22"
            required="required"
            type="date"
          />
          <!---->
          <!---->
          <!---->
        </div>
      </div>
       
      <button
        class="btn btn-primary"
        type="submit"
      >
        Add sprints
      </button>
       
      <button
        class="btn btn-grey"
        type="button"
      >
        Cancel
      </button>
    </div>
     
    <div
      class="col-md-6"
    >
      <h4>
        Preview
      </h4>
       
      <div>
        <strong>
          1. Sprint:
        </strong>
        

                    2222-02-22
                    -
                    2222-02-23
                
      </div>
      <div>
        <strong>
          2. Sprint:
        </strong>
        

                    2222-02-24
                    -
                    2222-02-25
                
      </div>
      <div>
        <strong>
          3. Sprint:
        </strong>
        

                    2222-02-26
                    -
                    2222-02-27
                
      </div>
    </div>
  </div>
</form>
`;
